File: HackerRank Solutions/Interview Preparation Kit/String Manipulation/Common Child.cpp
#include <bits/stdc++.h>

using namespace std;
static int T[5001][5001];
// Complete the commonChild function below.
int commonChild(string s1, string s2) {
    for(int i=0; i<=s1.size(); i++){
        for(int j=0; j<=s2.size(); j++){
            if(i == 0 || j == 0){
                T[i][j] = 0;
            }else if(s1[i-1] == s2[j-1]){
                T[i][j] = 1+T[i-1][j-1];
            }else{
                T[i][j] = max(T[i][j-1], T[i-1][j]);
            }
        }
    }
    return T[s1.size()][s2.size()];
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string s1;
    getline(cin, s1);

    string s2;
    getline(cin, s2);

    int result = commonChild(s1, s2);

    fout << result << "\n";

    fout.close();

    return 0;
}

File: HackerRank Solutions/Interview Preparation Kit/String Manipulation/Common Child.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {
    // Complete the commonChild function below.
    static int commonChild(String s1, String s2) {
        int L[][] = new int[s1.length()+1][s2.length()+1];
        for(int i=0;i<=s1.length();i++){
            for(int j=0;j<=s2.length();j++){
                if(i==0 || j==0)
                    L[i][j]=0;
                else if(s1.charAt(i-1)==s2.charAt(j-1)){
                    L[i][j] = L[i-1][j-1]+1;
                }
                else{
                    L[i][j] = Math.max(L[i-1][j],L[i][j-1]);
                }
            }
        }
        return L[s1.length()][s2.length()];
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String s1 = scanner.nextLine();

        String s2 = scanner.nextLine();

        int result = commonChild(s1, s2);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedWriter.close();

        scanner.close();
    }
}
